### PR TITLE
consult healtcheck probes when publishing enpoints 

### PR DESCRIPTION
**Pull request by maksymvavilov**
Do not publish unhealthy endpoints or remove them from provider

### CODE REVIEW

File: internal/controller/dnsrecord_healthchecks.go [hunk 1]
```diff
@@ -94,6 +95,69 @@ func (r *DNSRecordReconciler) ensureProbe(ctx context.Context, generated *v1alph
 	return nil
 }
 
+// removeUnhealthyEndpoints fetches all probes associated with this record and uses the following criteria while removing endpoints:
+//   - If the Leaf Address has no health check CR - it is healthy
+//   - If the health check CR has insufficient failures - it is healthy
+//   - If the health check CR is deleting - it is healthy
+//   - If the health check is a CNAME and any IP is healthy - the CNAME is healthy
```
Comment (maleck13) [inline internal/controller/dnsrecord_healthchecks.go:102]: Is it also the case that if it is the last endpoint it healthy?

File: internal/controller/dnsrecord_healthchecks.go [hunk 2]
```diff
@@ -94,6 +95,69 @@ func (r *DNSRecordReconciler) ensureProbe(ctx context.Context, generated *v1alph
...
+func (r *DNSRecordReconciler) removeUnhealthyEndpoints(ctx context.Context, endpoints []*endpoint.Endpoint, dnsRecord *v1alpha1.DNSRecord) ([]*endpoint.Endpoint, bool, error) {
```
Comment (maleck13) [inline internal/controller/dnsrecord_healthchecks.go:103]: what is the boolean for?

Comment (maksymvavilov) [reply internal/controller/dnsrecord_healthchecks.go:103]: for reporting to status. Changed to string slice for more information in the next PR. I decided to add it here so the next PR will not be changing this file, but I ended up refactoring it anyway. Could remove for this one if this matters

File: internal/controller/dnsrecord_healthchecks.go (hunk 1 above)
Comment (maksymvavilov) [reply internal/controller/dnsrecord_healthchecks.go:102]: Nope, if this would remove all the enpoints we will do nothing. I'll add more comments

Comment (maleck13) [reply internal/controller/dnsrecord_healthchecks.go:102]: Discussed. We wont remove records only will remove if multi-value record

File: internal/controller/dnsrecord_controller_test.go [hunk 3]
```diff
@@ -1,4 +1,4 @@
-//go:build integration
+//go:build integration2
```
Comment (maleck13) [inline internal/controller/dnsrecord_controller_test.go:1]: what is this for? Is something you setup for local use?

File: internal/controller/dnsrecord_healthchecks.go [hunk 4]
```diff
@@ -94,6 +95,72 @@ func (r *DNSRecordReconciler) ensureProbe(ctx context.Context, generated *v1alph
...
 
+// removeUnhealthyEndpoints fetches all probes associated with this record and uses the following criteria while removing endpoints:
+//   - If the Leaf Address has no health check CR - it is healthy
+//   - If the health check CR has insufficient failures - it is healthy
+//   - If the health check CR is deleting - it is healthy
+//   - If the health check is a CNAME and any IP is healthy - the CNAME is healthy
+//
+// If this leads to an empty array of endpoints it:
+//   - Does nothing (prevents NXDomain response) if we already published
+//   - Returns empty array of nothing is published (prevent from publishing unhealthy EPs)
```
Comment (maleck13) [inline internal/controller/dnsrecord_healthchecks.go:106]:
```suggestion
//   - Returns empty array if nothing is published (prevent from publishing unhealthy EPs)
```

File: internal/controller/dnsrecord_healthchecks.go [hunk 5]
```diff
@@ -94,6 +95,72 @@ func (r *DNSRecordReconciler) ensureProbe(ctx context.Context, generated *v1alph
...
+//
+// it returns the list of healthy endpoints, an array of unhealthy addresses and an error
+func (r *DNSRecordReconciler) removeUnhealthyEndpoints(ctx context.Context, specEndpoints []*endpoint.Endpoint, dnsRecord *v1alpha1.DNSRecord) ([]*endpoint.Endpoint, []string, error) {
```
Comment (maleck13) [inline internal/controller/dnsrecord_healthchecks.go:109]: We don't seem to be using the second return []string? Should we just remove it?

File: internal/controller/dnsrecord_healthchecks.go [hunk 6]
```diff
@@ -94,6 +95,72 @@ func (r *DNSRecordReconciler) ensureProbe(ctx context.Context, generated *v1alph
...
+			tree.RemoveNode(&common.DNSTreeNode{
+				Name: probe.Spec.Address,
+			})
+			unhealthyAddresses = append(unhealthyAddresses, probe.Spec.Address)
+		}
+
+	}
+
+	// if at least one of the leaf probes was healthy return healthy probes
+	if haveHealthyProbes {
```
Comment (maleck13) [inline internal/controller/dnsrecord_healthchecks.go:157]:
```suggestion
	if isHealthy {
```

File: internal/controller/dnsrecord_healthchecks.go [hunk 7]
```diff
@@ -94,6 +95,72 @@ func (r *DNSRecordReconciler) ensureProbe(ctx context.Context, generated *v1alph
...
+		return *common.ToEndpoints(tree, ptr.To([]*endpoint.Endpoint{})), unhealthyAddresses, nil
+	}
+	// if none of the probes are healthy or probes don't exist - don't modify endpoints
+	return dnsRecord.Status.Endpoints, unhealthyAddresses, nil
```
Comment (maleck13) [inline internal/controller/dnsrecord_healthchecks.go:161]: right so this is where, if we had an unhealthy record, we would want to validate that there would still be at least one record returned for a possible dns path (IE for a given geo record set or across all record sets (if no geo loadbalancing).

File: internal/controller/dnsrecord_healthchecks.go [hunk 8]
```diff
@@ -94,6 +95,72 @@ func (r *DNSRecordReconciler) ensureProbe(ctx context.Context, generated *v1alph
...
+			})
+			unhealthyAddresses = append(unhealthyAddresses, probe.Spec.Address)
+		}
+
+	}
+
+	// if at least one of the leaf probes was healthy return healthy probes
+	if haveHealthyProbes {
+		return *common.ToEndpoints(tree, ptr.To([]*endpoint.Endpoint{})), unhealthyAddresses, nil
+	}
```
Comment (maleck13) [inline internal/controller/dnsrecord_healthchecks.go:159]: is there value here of adding a status condition to the DNSRecord?
```
if !isHealthy{
conditions.AddCondition(...
{
type: Healthy
Reason: NoHealthChecksPassing
Status: False
Message: Record is unhealthy 
})```
}

**Review by maleck13 — COMMENTED** [review]
This looks pretty good. I will test it some more tomorrow

**Pull request merged**